src/sgb-xi/index.ts

```diff
@@ -77,13 +97,17 @@ const forEachKostentraeger = (invoices: Invoice[], rechnungsart: RechnungsartSch
 const groupFaelleByKostentraeger = (faelle: Abrechnungsfall[]) =>
     valuesGroupedBy(faelle, fall => fall.versicherter.kostentraegerIK);
 
-const groupByLeistungsart = (faelle: Abrechnungsfall[]) => faelle
-    .map(fall => {
-        return fall.einsaetze.map(einsatz =>
-            valuesGroupedBy(einsatz.leistungen, leistung => leistung.leistungsart)
-                .map(leistungen => ({ ...einsatz, leistungen } as Einsatz))
-        ).map(einsaetze => ({ ...fall, einsaetze } as Abrechnungsfall))
-    });
+const groupByLeistungsart = (faelle: Abrechnungsfall[]) => faelle.flatMap(fall => [
+    [...new Set(
+        fall.einsaetze.flatMap(einsatz => einsatz.leistungen).map(leistung => leistung.leistungsart)
+    )].map(leistungsart => ({
+        ...fall,
+        einsaetze: fall.einsaetze.map(einsatz => ({
+            ...einsatz,
+            leistungen: einsatz.leistungen.filter(leistung => leistung.leistungsart == leistungsart)
+        })).filter(einsatz => einsatz.leistungen.length)
+    }))
+]);
```

Comment:
Just out of curiousity / doubting my own abilities - when you read this again now, you find this code easily readable and understandable?

Maybe I just have to get used to this functional kind of programming but what I am missing most are some intermediate variables with names that explain in what kind of format the data is now. Maybe a documentation comment and a type declaration of the result would also help.

Judging from the function name, does this do something like 
https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/associate-by.html
if yes, maybe a utility function that mimics this kotlin function would make the code more easily readable.

Reply:
Yes, this is complex code and I remember it took me a long time to get this right. As far as I understand from the documentation, associateBy is like the groupBy function in src/utils.ts. But groupByLeistungsart is different, because each Abrechnungsfall has to be split based on the leistungsart which is stored deep down the nested structure, because each invoice in the Nutzdaten file format has to be for one kind of leistungsart only. 

I added some documentation, result type and outsourced a part of the code to be more clear on what the function does and how. The format of the data does not change though, it just splits every Abrechnungsfall in multiple if needed. I don't see how the function can be simplified further without breaking things.